packages/opentelemetry-resource-util/src/detector/detector.ts

```diff
@@ -18,19 +18,19 @@ import {
   CloudProviderValues,
 } from '@opentelemetry/semantic-conventions';
 
-import {Detector, Resource} from '@opentelemetry/resources';
+import {Detector, DetectorSync, Resource} from '@opentelemetry/resources';
 import * as gce from './gce';
 import * as gke from './gke';
 import * as faas from './faas';
 import * as gae from './gae';
 import * as metadata from 'gcp-metadata';
 import {Attributes} from '@opentelemetry/api';
 
-export class GcpDetector implements Detector {
+export class GcpDetectorBase {
```

Comment:
Nit: I generally avoid using class inheritance to model shared implementation. How about leaving `GcpDetector` unchanged (with the exception of a deprecation comment), and writing `GcpDetectorSync` using composition:

```ts
export class GcpDetectorSync implements DetectorSync {
  private asyncDetector = new GcpDetector();

  detect(): Resource {
    return new Resource({}, asyncDetector().then(res => res.attributes));
  }
}
```

Reply:
I refactored the impl into a standalone function. The class was only being used for caching `projectId` which idek why I added, as it never gets re-read in the one-shot case.